File: packages/websearch/src/fence.ts
import { toolError, type ToolError } from "@agent-sh/harness-core";
import type {
  WebSearchSafeSearch,
  WebSearchSessionConfig,
  WebSearchTimeRange,
} from "./types.js";

/**
 * Permission hook call for websearch. Mirrors the shape used by webfetch /
 * read / grep / bash but with WebSearch-specific metadata. Permission is
 * keyed on the backend, not the query (you trust a backend, not individual
 * searches). Returns a decision string; "ask" is treated as "deny" in
 * autonomous mode.
 */
export async function askPermission(
  session: WebSearchSessionConfig,
  args: {
    query: string;
    backendUrl: string;
    backendHost: string;
    count: number;
    timeRange: WebSearchTimeRange;
    safeSearch: WebSearchSafeSearch;
    categories: readonly string[];
  },
): Promise<
  | { decision: "allow" | "allow_once" }
  | { decision: "deny"; reason: string }
> {
  const { permissions } = session;
  const pattern = `WebSearch(backend:${args.backendHost})`;

  if (permissions.hook === undefined) {
    if (permissions.unsafeAllowSearchWithoutHook === true) {
      return { decision: "allow" };
    }
    return {
      decision: "deny",
      reason:
        "websearch tool has no permission hook configured; refusing to query the search backend. Wire a hook or set session.permissions.unsafeAllowSearchWithoutHook for test fixtures.",
    };
  }

  // A search query is low-sensitivity and audit-useful, so it's logged —
  // unless the session opts to log only its length.
  const queryField = session.redactQueryInHook === true
    ? { query_length: args.query.length }
    : { query: args.query };

  const decision = await permissions.hook({
    tool: "websearch",
    path: args.backendUrl,
    action: "read",
    always_patterns: [pattern],
    metadata: {
      ...queryField,
      count: args.count,
      time_range: args.timeRange,
      safe_search: args.safeSearch,
      categories: args.categories,
      backend_host: args.backendHost,
    },
  });
  if (decision === "deny") {
    return {
      decision: "deny",
      reason: `Search blocked by permission policy. Pattern hint: ${pattern}`,
    };
  }
  if (decision === "allow" || decision === "allow_once") {
    return { decision };
  }
  return {
    decision: "deny",
    reason:
      "Permission hook returned 'ask' but websearch runs in autonomous mode. Configure the hook to return allow or deny.",
  };
}

export function permissionDeniedError(
  query: string,
  reason: string,
): ToolError {
  const echoQuery = query.length > 300 ? query.slice(0, 300) + "..." : query;
  return toolError(
    "PERMISSION_DENIED",
    `${reason}\nQuery: "${echoQuery}"`,
    { meta: { query } },
  );
}

File: crates/websearch/src/fence.rs
use harness_core::permissions::PermissionQuery;
use harness_core::{PermissionDecision, ToolError, ToolErrorCode};

use crate::types::{SafeSearch, WebSearchSessionConfig, WebSearchTimeRange};

#[derive(Debug, Clone)]
pub enum PermissionOutcome {
    Allow,
    Deny { reason: String },
}

pub struct AskArgs<'a> {
    pub query: &'a str,
    pub backend_url: &'a str,
    pub backend_host: &'a str,
    pub count: usize,
    pub time_range: WebSearchTimeRange,
    pub safe_search: SafeSearch,
    pub categories: &'a [String],
}

pub async fn ask_permission(
    session: &WebSearchSessionConfig,
    args: AskArgs<'_>,
) -> PermissionOutcome {
    let permissions = &session.permissions.inner;
    let pattern = format!("WebSearch(backend:{})", args.backend_host);

    if permissions.hook.is_none() {
        if session.permissions.unsafe_allow_search_without_hook {
            return PermissionOutcome::Allow;
        }
        return PermissionOutcome::Deny {
            reason: "websearch tool has no permission hook configured; refusing to query the search backend. Wire a hook or set permissions.unsafe_allow_search_without_hook for test fixtures."
                .to_string(),
        };
    }

    // A search query is low-sensitivity and audit-useful, so it's logged —
    // unless the session opts to log only its length.
    let mut metadata = serde_json::json!({
        "count": args.count,
        "time_range": args.time_range.as_str(),
        "safe_search": args.safe_search.as_str(),
        "categories": args.categories,
        "backend_host": args.backend_host,
    });
    if session.redact_query_in_hook {
        metadata["query_length"] = serde_json::json!(args.query.chars().count());
    } else {
        metadata["query"] = serde_json::json!(args.query);
    }

    let query = PermissionQuery {
        tool: "websearch".to_string(),
        path: args.backend_url.to_string(),
        action: "read".to_string(),
        always_patterns: vec![pattern.clone()],
        metadata,
    };
    let hook = permissions.hook.as_ref().unwrap();
    let decision = (hook)(query).await;
    match decision {
        PermissionDecision::Allow | PermissionDecision::AllowOnce => PermissionOutcome::Allow,
        PermissionDecision::Deny => PermissionOutcome::Deny {
            reason: format!(
                "Search blocked by permission policy. Pattern hint: {}",
                pattern
            ),
        },
        PermissionDecision::Ask => PermissionOutcome::Deny {
            reason: "Permission hook returned 'ask' but websearch runs in autonomous mode. Configure the hook to return allow or deny.".to_string(),
        },
    }
}

pub fn permission_denied_error(query: &str, reason: &str) -> ToolError {
    let echo_query = if query.chars().count() > 300 {
        let truncated: String = query.chars().take(300).collect();
        format!("{}...", truncated)
    } else {
        query.to_string()
    };
    ToolError::new(
        ToolErrorCode::PermissionDenied,
        format!("{}\nQuery: \"{}\"", reason, echo_query),
    )
    .with_meta(serde_json::json!({ "query": query }))
}
